compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt

```diff
@@ -96,19 +97,21 @@ class ControlFlowGraphBuilder {
 
     // ----------------------------------- Public API -----------------------------------
 
-    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirStatement>? {
+    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirExpression>? {
         val exitNode = function.controlFlowGraphReference?.controlFlowGraph?.exitNode ?: return null
 
-        fun FirElement.extractArgument(): FirElement = when {
-            this is FirReturnExpression && target.labeledElement.symbol == function.symbol -> result.extractArgument()
-            else -> this
-        }
-
-        fun CFGNode<*>.extractArgument(): FirStatement? = when (this) {
-            is FunctionEnterNode, is TryMainBlockEnterNode, is FinallyBlockExitNode, is CatchClauseEnterNode -> null
-            is BlockExitNode -> if (function.isLambda || isDead) firstPreviousNode.extractArgument() else null
-            is StubNode -> firstPreviousNode.extractArgument()
-            else -> fir.extractArgument() as FirStatement?
+        fun CFGNode<*>.returnExpression(): FirExpression? = when (this) {
+            is BlockExitNode -> when {
+                // lambda@{ x } -> x
+                // lambda@{ class C } -> Unit-returning stub
+                function.isLambda -> fir.statements.lastOrNull() as? FirExpression
+                    ?: buildUnitExpression { source = fir.statements.lastOrNull()?.source ?: fir.source }
```

Comment:
It's better to create some `FirStubUnitExpression` singleton without a source here

Reply:
Without a source there won't be an error when a lambda with an expected non-Unit return type ends in a non-expression. Probably one of the checkers is incorrect...